pkg/datapath/loader/loader.go

```diff
@@ -132,6 +137,63 @@ func nullifyStringSubstitutions(strings map[string]string) map[string]string {
 	return nullStrings
 }
 
+// patchHostNetdevHeaderFile creates a header file corresponding to each new
+// device to which bpf_host is attached, including the second host device and
+// native devices.
+func patchHostNetdevHeaderFile(stateDir, ifName string, mac mac.MAC, ifIndex uint32,
+	secctxFromIpcache uint32, nodePort bool) (netdevHeaderFile string, err error) {
+	netdevHeaderFilename := strings.Replace(common.CHeaderFileName, ".h", "_"+ifName+".h", 1)
+	netdevHeaderFile = path.Join(stateDir, netdevHeaderFilename)
+	headerFile := path.Join(stateDir, common.CHeaderFileName)
+
+	fin, err := os.Open(headerFile)
+	if err != nil {
+		return
+	}
+	defer fin.Close()
+	fout, err := os.Create(netdevHeaderFile)
+	if err != nil {
+		return
+	}
+	defer func() {
+		if err != nil {
+			os.Remove(netdevHeaderFile)
+		}
+	}()
+	defer fout.Close()
+
+	rd := bufio.NewReader(fin)
+	for {
+		line, err := rd.ReadString('\n')
+		if err != nil && err != io.EOF {
+			return netdevHeaderFile, err
+		}
```

Comment:
Any reason not to use a `bufio.Scanner` here, which would simplify this logic a bit?

```go
scanner := bufio.NewScanner(fin)
for scanner.Scan() {
	line := scanner.Text()
	// line now contains the line without trailing \n
	...
}

if scanner.Err() != nil {
	return nil, scanner.Err()
}
```

Reply:
I'm wondering why not take it a little further.

All the state for this file should be in-memory before it lands on disk. Rather than writing one version to disk, then going to the next host endpoint prog, reading in the data and trying to substitute text & format correctly, why not copy the structure in-memory, fix it up, and just serialize to disk?

Reply:
@joestringer I'm not sure I understand your suggestion. Are you saying we should instead patch the object files and generate the header files during the compilation step instead of doing it during the loading step?

I agree that the dependence on the header file's format is far from ideal. Longer term, I think the correct way to fix this is to have support for multiple BPF programs per endpoint in the agent. We need it for https://github.com/cilium/cilium/issues/12604 anyway. Once we can have several BPF programs per endpoint, we won't need any special handling for the native device's programs anymore.

Reply:
> I'm not sure I understand your suggestion. Are you saying we should instead patch the object files and generate the header files during the compilation step instead of doing it during the loading step?

My core (hopefully constructive?) criticism is that this code is reading a headerfile and I don't understand why we need to read the headerfile to write state to disk. The reading portion seems unnecessarily complicated, why not just write the unique values to a dedicated headerfile?

I agree there are longer term steps we will want to take steps towards, my hope in bringing up topics like this is to minimize extra code to rework later on :-)

Reply:
In essence what I had in mind was that we find a `datapath.ConfigWriter` from somewhere and we pass the `datapath.Endpoint` here into the `ConfigWriter.WriteEndpointConfig` and it generates this file instead of reading & substituting values. To achieve this, I think yes we would need to move specifically the headerfile writing earlier just to gain access to the ConfigWriter (or pass the ConfigWriter through these paths, but I don't think that makes much sense).

Traditionally this would be a requirement since we used the headerfiles to compile the program so you would *have* to write the header first and then fix up the ELF later but I understand that this is no longer a strict requirement.

Reply:
My main concern here is whether reading the state from the filesystem makes this code path more brittle; garbage-in -> garbage out type issues, or having a different code path here compared to the "main" path that generally generates the headerfile, which over time can lead to the different paths getting out of sync, or there's a bug that affects one path and not the other, those sorts of things.

Maybe a little bit of my concern here is also that it seems like we may be adding code here that we know we will want to rework later on, but by doing so we further entangle the codebase, adding one more thing to the growing list of things we need to consider to tackle the "longer term" goals that you already allude to. I get the feeling there are other chickens like this coming home to roost already.

That said, I haven't scoped out what I'm asking for here so maybe what I'm really thinking about is some variation on the longer-term fix and it's a lot more work. In that case, maybe it's best to merge this as-is and tackle such topics going forward.